### ISSUE TITLE
Required property within array ommitted

### ISSUE THREAD

**Issue by michaelmior** [json-schema-faker/json-schema-faker#747]
Schema

```json
{
  "properties": {
    "foo": {
      "items": {
        "properties": {
          "bar": {
            "type": "null"
          },
          "baz": {
            "properties": {
              "quux": {
                "type": "string"
              },
              "corge": {
                "type": "number"
              }
            },
            "required": [],
            "type": "object",
            "additionalProperties": false
          }
        },
        "required": [
          "bar",
          "baz"
        ],
        "type": "object"
      },
      "type": "array"
    }
  },
  "required": [
    "foo"
  ],
  "type": "object"
}
```

Generated output

```json
{
  "foo": [
    {
      "bar": null
    }
  ]
}
```

Note that the second array element generated does not have the required property `baz`. I've tried simplifying the schema in several ways. When I make any of the following changes, I don't seem to have the same issue. I've also tested multiple versions as far back as `v0.5.0-rc1` and the issue appears to be present in all the older versions I have tested as well.

1. Remove the array and generate simple objects.
2. Remove the property `bar`.
3. Remove either `corge` or `quux` so that `baz` only has a single nested property.

**Comment by pateketrueke** [comment]
Interesting, I tested it on the webapp (0.5.0-rcv.45) and seems to be working, also I tested it locally (0.5.0-rcv.46) and I can see `baz` values generated.

Do you have a working repro to investigate? Thank you!

**Comment by michaelmior** [comment]
@pateketrueke I tested in the web app, and I got the same bug. It doesn't happen all the time though. But pasting in the exact schema above, it only takes a handful of tries before I get at least one array element with `baz` missing.

**Comment by pateketrueke** [comment]
Got your point now, `buz` is not always generated because their properties are not always required (`"required": []`), sometimes it'll generate an empty object and the library just remove those empty objects.

However that's not what you expected, what is the behavior you're expecting then?

**Comment by michaelmior** [comment]
I figured that's what's happening, but if I understand the JSON Schema specification correctly, a required property *must* be specified, even if it's an empty object. So I would expect the empty object to be preserved if the property is required.

**Comment by pateketrueke** [comment]
You're right! I figured out why those empty objects are not there, and effectively it's a bug.

We have a `shouldClean()` function that checks if a generated property can be removed from the schema if empty, and we have been reading the schema as-is but in this case, we should use the `schema.items` schema to read the `required` property.

I've already fixed this issue on the `develop` branch if you want to verify. 🙏🏽

**Comment by pateketrueke** [comment]
> I figured that's what's happening, but if I understand the JSON Schema specification correctly, a required property must be specified, even if it's an empty object. So I would expect the empty object to be preserved if the property is required.

I validated the schema against https://www.jsonschemavalidator.net/ and seems like `required` properties should not be empty.

**Comment by michaelmior** [comment]
@pateketrueke That may be correct, but leaving the `required` property out altogether should have the same result.

**Comment by michaelmior** [comment]
The [JSON Schema website](https://json-schema.org/understanding-json-schema/reference/object.html#required-properties) states the following:

> The `required` keyword takes an array of zero or more strings. Each of these strings must be unique.

Note that it explicitly says *zero* or more strings.

**Comment by pateketrueke** [comment]
You're right again! That's already fixed, no problem. :beers:

**Comment by michaelmior** [comment]
@pateketrueke Thanks for fixing so quickly!

**Comment by pateketrueke** [comment]
> Note that it explicitly says zero or more strings.

Yeah, don't worry about that. I was trying to add a test-case for this, and got `[z-schema] Keyword 'required' must be an array with at least one element` so I am hesitating a bit how to cover this as well.

**Comment by michaelmior** [comment]
@pateketrueke I think it's reasonable to assume that the input to json-schema-faker is a valid schema. If an empty `required` array behaves the same as having `required` unspecified, I don't really see why anyone would have problem. If someone thinks their schema should not allow an empty `required` array, they should use some other schema validator that enforces that.

**Comment by Ceres6** [comment]
Hi, I'm having the same issue and maybe my slightly simplified version can come in handy here
```{
  "type": "object",
  "properties": {
    "d": {
      "maxItems": 1,
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "c": {
            "type": "object",
            "required": [
              "b"
            ],
            "properties": {
              "b": {
                "type": "object",
                "properties": {
                  "a": {
                    "type": "string"
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
``` 

if a reproducible example is needed 
```js
import { JSONSchemaFaker } from 'json-schema-faker'
import seedrandom from 'seedrandom'

JSONSchemaFaker.option({
  fillProperties: false,
  // random: seedrandom(seed)
  random: seedrandom('70ffb365')
})

const schema = {
  type: 'object',
  properties: {
    d: {
      maxItems: 1,
      type: 'array',
      items: {
        type: 'object',
        properties: {
          c: {
            type: 'object',
            required: ['b'],
            properties: {
              b: {
                type: 'object',
                properties: {
                  a: {
                    type: 'string'
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

const fake = JSONSchemaFaker.generate(schema)
```
If any work is needed on this I'd be glad to help

**Comment by l0kal** [comment]
I have the same issue when validating the generated JSON against the schema with AJV. I also tried virtually every combination of options I thought was relevant. Right now I am using `alwaysFakeOptionals: true,` which will give me the correct (at least what I deem as correct) result roughly 85% of the time.

my schema: https://github.com/extenda/hiiretail-json-schema-registry/blob/master/customer-config/checkout-app/cha.settings.v1.json

The error I see is
`{
        error: '"server" property must have required property "host"',
        path: '/pos/server'
      }`
      
The `host` property is required inside an optional property `server` this is defined in `"$ref": "#/$defs/ServerType"`.
As far as I understand the `host` and `networkCallTimeout` should be generated with fake values if the server object was faked.
Any advice on what options to set in the faker is appreciated :smile:
